extensions/authentication-passport/src/__tests__/acceptance/authentication-with-passport-strategy-oauth2-adapter.acceptance.ts

```diff
@@ -0,0 +1,194 @@
+ * 
+ * 2. thirdPartyCallBack
+ *           i. this is the callback for the thirdparty app
+ *          ii. on successful user login the third party calls this endpoint with an access code
+ *         iii. the passport oauth2 strategy exchanges the code for an access token
+ *          iv. the passport oauth2 strategy then calls the provided `verify()` function with the access token
+ */
+export class Oauth2Controller {
+  constructor() {}
+
+  // this configures the oauth2 strategy
+  @authenticate('oauth2')
+  // we have modeled this as a GET endpoint
+  @get('/auth/thirdparty')
+  // loginToThirdParty() is the handler for '/auth/thirdparty'
+  // this method is injected with 'x-loopback-authentication-redirect-url'
+  // the value for 'x-loopback-authentication-redirect-url' is set by the passport strategy adapter
+  loginToThirdParty(@param.query.string('x-loopback-authentication-redirect-url') redirectUrl: string,
+  @param.query.number('x-loopback-authentication-redirect-status') status: number) {
+    return new RedirectRoute('/', redirectUrl, status);
+  }
+
+  // we configure the callback url also with the same oauth2 strategy
+  @authenticate('oauth2')
+  // this SHOULD be a GET call so that the third party can redirect
+  @get('/auth/thirdparty/callback')
+  // thirdPartyCallBack() is the handler for '/auth/thirdparty/callback'
+  // the oauth2 strategy identifies this as a callback with the request.query.code sent by the third party app
+  // the oauth2 strategy exchanges the access code for a access token and then calls the provided verify() function
+  // the verify function creates a user profile after verifying the access token
+  thirdPartyCallBack(@inject(SecurityBindings.USER) user: UserProfile) {
+    console.log('thirdPartyCallBack', user);
+    return user.token;
+  }
+}
+
+describe.only('Oauth2 authorization flow', () => {
+  let app: RestApplication;
+  let oauth2Strategy: StrategyAdapter<MyUser>;
+  let client: Client;
+
+  before(startMockProvider);
+  after(stopMockProvider);
+
+  before(givenLoopBackApp);
+  before(givenOauth2Strategy);
+  before(setupAuthentication);
+  before(givenControllerInApp);
+  before(givenClient);
+
+  let oauthProviderUrl: string;
+  let providerLoginUrl: string;
+  let callbackToLbApp: string;
+
+  context('when client invokes oauth flow', () => {
+
+    it('call is redirected to third party authorization url', async () => {
+      const response = await client.get('/auth/thirdparty').expect(303);
+      oauthProviderUrl = response.get('Location');
+      expect(url.parse(response.get('Location')).pathname).to.equal(url.parse(oauth2Options.authorizationURL).pathname);
+    });
+
+    it('call to authorization url is redirected to oauth providers login page', async () => {
+      const response = await supertest('').get(oauthProviderUrl).expect(302);
+      providerLoginUrl = response.get('Location');
+      expect(url.parse(response.get('Location')).pathname).to.equal('/login');
```

Comment:
So to be clear, you are running an http server that is capable of hosting these urls, but we are only testing the calling of these endpoints, but obviously not interacting with the html fields (user and password and submit).

Reply:
the test for login is also done , please see 
https://github.com/strongloop/loopback-next/blob/75a02cabe61be1414a22084984ec87e1982ee17c/extensions/authentication-passport/src/__tests__/acceptance/authentication-with-passport-strategy-oauth2-adapter.acceptance.ts#L148-L152

Reply:
the entire oauth2 auth flow is tested, we only test deliberately for redirected urls, because browsers handle that redirection when seeing a 302 or 303.

Reply:
I see you are mocking that the form was filled by user by providing the parameters in the url. But there is no UI robot clicking and typing...is what I meant.